os/bluestore: Introduce zoned freelist manager.

agayev opened the pull request:
os/bluestore: Introduce zoned freelist manager.

This patch introduced ZonedFreelistManager, which tracks the write pointer and
the number of dead bytes per zone.

This patch depends on https://github.com/ceph/ceph/pull/35294

Signed-off-by: Abutalib Aghayev <agayev@cs.cmu.edu>

tchaikov commented:
```
2020-06-12 03:00:09,487.487 INFO:__main__:test_standby (tasks.mgr.test_dashboard.TestDashboard) ... ERROR
2020-06-12 03:00:09,488.488 INFO:__main__:Stopped test: test_standby (tasks.mgr.test_dashboard.TestDashboard) in 51.20297s
2020-06-12 03:00:09,488.488 INFO:__main__:
2020-06-12 03:00:09,488.488 INFO:__main__:======================================================================
2020-06-12 03:00:09,489.489 INFO:__main__:ERROR: test_standby (tasks.mgr.test_dashboard.TestDashboard)
2020-06-12 03:00:09,489.489 INFO:__main__:----------------------------------------------------------------------
2020-06-12 03:00:09,489.489 INFO:__main__:Traceback (most recent call last):
2020-06-12 03:00:09,489.489 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connection.py", line 159, in _new_conn
2020-06-12 03:00:09,489.489 INFO:__main__:    (self._dns_host, self.port), self.timeout, **extra_kw)
2020-06-12 03:00:09,489.489 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/util/connection.py", line 80, in create_connection
2020-06-12 03:00:09,490.490 INFO:__main__:    raise err
2020-06-12 03:00:09,490.490 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/util/connection.py", line 70, in create_connection
2020-06-12 03:00:09,490.490 INFO:__main__:    sock.connect(sa)
2020-06-12 03:00:09,490.490 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/gevent/_socket3.py", line 407, in connect
2020-06-12 03:00:09,490.490 INFO:__main__:    raise error(err, strerror(err))
2020-06-12 03:00:09,490.490 INFO:__main__:ConnectionRefusedError: [Errno 111] Connection refused
2020-06-12 03:00:09,490.490 INFO:__main__:
2020-06-12 03:00:09,490.490 INFO:__main__:During handling of the above exception, another exception occurred:
2020-06-12 03:00:09,491.491 INFO:__main__:
2020-06-12 03:00:09,491.491 INFO:__main__:Traceback (most recent call last):
2020-06-12 03:00:09,491.491 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
2020-06-12 03:00:09,491.491 INFO:__main__:    chunked=chunked)
2020-06-12 03:00:09,491.491 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 343, in _make_request
2020-06-12 03:00:09,491.491 INFO:__main__:    self._validate_conn(conn)
2020-06-12 03:00:09,491.491 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 839, in _validate_conn
2020-06-12 03:00:09,492.492 INFO:__main__:    conn.connect()
2020-06-12 03:00:09,492.492 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connection.py", line 301, in connect
2020-06-12 03:00:09,492.492 INFO:__main__:    conn = self._new_conn()
2020-06-12 03:00:09,492.492 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connection.py", line 168, in _new_conn
2020-06-12 03:00:09,492.492 INFO:__main__:    self, "Failed to establish a new connection: %s" % e)
2020-06-12 03:00:09,492.492 INFO:__main__:urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x7fefa3db4be0>: Failed to establish a new connection: [Errno 111] Connection refused
2020-06-12 03:00:09,492.492 INFO:__main__:
2020-06-12 03:00:09,493.493 INFO:__main__:During handling of the above exception, another exception occurred:
2020-06-12 03:00:09,493.493 INFO:__main__:
2020-06-12 03:00:09,493.493 INFO:__main__:Traceback (most recent call last):
2020-06-12 03:00:09,493.493 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
2020-06-12 03:00:09,493.493 INFO:__main__:    timeout=timeout
2020-06-12 03:00:09,493.493 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 638, in urlopen
2020-06-12 03:00:09,493.493 INFO:__main__:    _stacktrace=sys.exc_info()[2])
2020-06-12 03:00:09,493.493 INFO:__main__:  File "/tmp/tmp.7s9dORHSKS/venv/lib/python3.6/site-packages/urllib3/util/retry.py", line 399, in increment
2020-06-12 03:00:09,494.494 INFO:__main__:    raise MaxRetryError(_pool, url, error or ResponseError(cause))
2020-06-12 03:00:09,494.494 INFO:__main__:urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='slave-ubuntu11.front.sepia.ceph.com', port=7789): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fefa3db4be0>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

tchaikov commented:
jenkins test dashboard backend

agayev commented:
> jenkins test dashboard backend

Do you mean these should pass as well?  If yes, how can I reproduce them locally?  It appears to be an unrelated to the changes I have made.  (Although I did specify -DWITH_MGR_DASHBOARD_FRONTEND to do_cmake.sh.)

tchaikov commented:
> > jenkins test dashboard backend
> 
> Do you mean these should pass as well? If yes, how can I reproduce them locally? It appears to be an unrelated to the changes I have made. (Although I did specify -DWITH_MGR_DASHBOARD_FRONTEND to do_cmake.sh.)

@agayev it's a trigger phrase to re-trigger the failed test. i wanted to understand if this failure was reproducible or not. if it is, it's much simpler for us to fix it.

and yes, it is indeed an unrelated failure to your changeset. so there is no need to take any action on your end. =)

agayev commented:
@tchaikov @ifed01 I messed up rebasing my PRs when syncing my tree.  Please ignore this PR and https://github.com/ceph/ceph/pull/35659 and review https://github.com/ceph/ceph/pull/35730 and https://github.com/ceph/ceph/pull/35731 instead.

Apologies for the mess up.  I will delete this and https://github.com/ceph/ceph/pull/35659 once you ack.

tchaikov commented:
> @tchaikov @ifed01 I messed up rebasing my PRs when syncing my tree. Please ignore this PR and #35659 and review #35730 and #35731 instead.
> 
> Apologies for the mess up. I will delete this and #35659 once you ack.

no worries. but i think you can always `git push --force` to your remote branch.

agayev commented:
> > @tchaikov @ifed01 I messed up rebasing my PRs when syncing my tree. Please ignore this PR and #35659 and review #35730 and #35731 instead.
> > Apologies for the mess up. I will delete this and #35659 once you ack.
> 
> no worries. but i think you can always `git push --force` to your remote branch.

@tchaikov I did that, but the github interface was showing some conflicts.  They are gone now.  Never mind, I will delete the new PRs.  @ifed01 please review this and https://github.com/ceph/ceph/pull/35659 when you get a chance.  Thanks!

ifed01 commented on src/os/bluestore/FreelistManager.cc:
```diff
@@ -16,10 +19,28 @@ FreelistManager *FreelistManager::create(
 }
 
-void FreelistManager::setup_merge_operators(KeyValueDB *db)
+void FreelistManager::setup_merge_operators(KeyValueDB *db,
```
Let's move this to virtual functions defined at specific FreeListManager implementations... For the sake of better encapsulations....

ifed01 replied:
never mind, we don't have fm instance at this point

ifed01 commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+
+std::vector<zone_state_t> ZonedFreelistManager::get_zone_states(
+    KeyValueDB *kvdb) const {
+  std::lock_guard l(lock);
```
Not sure one needs a lock here

agayev replied:
Will remove.

ifed01 commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+  int r = _read_cfg(cfg_reader);
+  if (r != 0) {
+    dout(1) << __func__ << " fall back to legacy meta repo" << dendl;
+    _load_from_db(kvdb);
```
No need to support loading from DB as there are no legacy deployments for zoned drives. Hence reading from  CFG is enough.

agayev replied:
Okay, will remove.

ifed01 commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+    uint64_t granularity,
+    KeyValueDB::Transaction txn) {
+  size = new_size;
+  bytes_per_block = granularity & 0x00000000ffffffff;
```
Please comment about piggy-backing here....

agayev replied:
Will add.

ifed01 commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+  } else {
+    enumerate_p->next();
+    if (!enumerate_p->valid()) {
+      dout(10) << __func__ << " end" << dendl;
```
dout(10) is IMO too verbose

agayev replied:
Will change it to dout(30).

ifed01 commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+  // starting case
+  if (enumerate_zone_num == ~0UL) {
+    enumerate_p = kvdb->get_iterator(info_prefix);
+    enumerate_p->lower_bound(string());
```
Check iterator validity?

agayev replied:
Okay.

ifed01 commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+  return 0;
+}
+
+void ZonedFreelistManager::_load_from_db(KeyValueDB* kvdb) {
```
no need for this func as per comment about lack of legacy deployments above

agayev replied:
Will remove.

ifed01 commented:
@agayev - you might want to rebase to get rid of the first commit

ifed01 commented on src/os/bluestore/zoned_types.h:
```diff
@@ -0,0 +1,84 @@
+    ceph_assert(v[1] <= zone_size);
+  }
+
+  void persist(const std::string& prefix, KeyValueDB::Transaction txn) {
```
You might want to have our regular encode/decode mechanics that maintain version compatibility  and some other stuff. To say nothing about the benefit of having uniform approach... ;)

agayev replied:
Okay, let me take a look at those.

agayev commented on src/os/bluestore/ZonedFreelistManager.cc:
```diff
@@ -0,0 +1,325 @@
+  std::vector<zone_state_t> zone_states;
+  auto p = kvdb->get_iterator(info_prefix);
+  for (p->lower_bound(string()); p->valid(); p->next()) {
+    zone_states.push_back(zone_state_t(p, zone_size));
```
Are you suggesting get_zone_states to take as an argument the range of zones to load?  Regarding dependence, I don't see how zone_states_t is allocator-dependent.

Perhaps you can sketch out what you have in mind, and I will implement it.

ifed01 replied:
I meant something like passing a callback (targeting to allocator) which to be called when a portion of zones is loaded.
But honestly I expected much smaller zones (something like ten MBs) and hence much longer zone list....

agayev replied:
If you don't mind, I will leave it as is for now because there is only one client of get_zone_states(), which runs at startup -- i.e. it is not on a critical path.  We can revisit later if need be.

ifed01 replied:
sounds good.

agayev commented on src/os/bluestore/zoned_types.h:
```diff
@@ -0,0 +1,84 @@
+struct zone_state_t {
+  uint64_t v[2];
+  uint64_t zone_num;
+  uint64_t zone_size;
```
So zoned devices have zone size and zone capacity associated with each zone.  Zone size is fixed and it is some multiple of 2 and it is used to quickly figure out in which zone a given LBA falls.  Zone capacity, on the other hand, is the actual capacity of a zone and it determines how much data you can actually store in a zone.  With HM-SMR HDDs, zone size == zone capacity, but with ZNS SSDs, zone size >= zone capacity, and furthermore, zone capacity varies from zone to zone due to NAND flash characteristics.  So we do need to keep zone capacity per zone, but I will change the name from zone_size to zone_capacity because that is what it is.

agayev replied:
Ended up just removing zone_size, for now.

agayev commented on src/os/bluestore/zoned_types.h:
```diff
@@ -0,0 +1,84 @@
+// state.  Calling persist method inserts the in-memory state to the database.
+struct zone_state_t {
+  uint64_t v[2];
+  uint64_t zone_num;
```
Sounds good.

agayev commented on src/os/bluestore/zoned_types.h:
```diff
@@ -0,0 +1,84 @@
+// We use the same struct for an on-disk and in-memory representation of the
+// state.  Calling persist method inserts the in-memory state to the database.
+struct zone_state_t {
+  uint64_t v[2];
```
Currently, ZNS SSDs have 1GB zones.  I wouldn't be surprised if they go to 4GB within a few years, but never mind, I will change it to uint32_t.

agayev commented:
@ifed01 I think I addressed all the issues you raised, except the one that I did not mark "resolved".  You can either elaborate on that and I can see what I can do, or you we can skip it for now.  Please take a look when you get a chance.  Thanks!

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -5094,6 +5113,11 @@ int BlueStore::_open_alloc()
   }
 
+  if (bdev->is_smr()) {
+    min_alloc_size &= 0x00000000ffffffff;
```
You might want to remove similar stuff reverting min_alloc_size immediately after Allocator::create call.
Or just omit min_alloc_size spoiling and use another local variable when passing to Allocator::create. Hence no need to revert...

agayev replied:
Done.  Used another variable and moved it to a function to avoid duplication.

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -6454,6 +6478,10 @@ int BlueStore::mkfs()
   if (r < 0)
     goto out_close_fsid;
 
+  if (bdev->is_smr()) {
```
May be better do that inside _open_bdev?

agayev replied:
Done.

ifed01 commented on src/os/bluestore/ZonedAllocator.cc:
```diff
@@ -27,13 +28,12 @@ ZonedAllocator::ZonedAllocator(CephContext* cct,
-      starting_zone((block_size & 0xffff000000000000) >> 48),
-      nr_zones(size / zone_size),
-      write_pointers(nr_zones) {
+      starting_zone_num((block_size & 0xffff000000000000) >> 48),
```
leave a comment about blocksize piggybacking?

agayev replied:
Done.  Thanks for the review and the approval!

agayev commented:
jenkins make check

agayev commented:
jenkins test make check

agayev commented:
jenkins test make check

agayev commented:
@tchaikov this can be tested and merged now.  Thanks!

tchaikov commented:
- https://pulpito.ceph.com/kchai-2020-07-06_04:02:43-rados-wip-kefu-testing-2020-07-06-1016-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-06_11:39:50-rados-wip-kefu-testing-2020-07-06-1016-distro-basic-smithi/
